### PR TITLE
Add ssl verification meta data

### PR DESCRIPTION
**Pull request by mbenita-Cyberark**
### Desired Outcome

Implement the full flow of working with trust store e2e

### Implemented Changes

TBD

### Connected Issue/Story

Resolves #[relevant GitHub issue(s), e.g. 76]

CyberArk internal issue link: [insert issue ID]()

### Definition of Done
*At least 1 todo must be completed in the sections below for the PR to be
merged.*

#### Changelog

- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a
  CHANGELOG update

#### Test coverage

- [ ] This PR includes new unit and integration tests to go with the code
  changes, or
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [ ] A follow-up issue to update official docs has been filed here: [insert issue ID]()
- [ ] This PR does not require updating any documentation

#### Behavior

- [ ] This PR changes product behavior and has been reviewed by a PO, or
- [ ] These changes are part of a larger initiative that will be reviewed later, or
- [ ] No behavior was changed with this PR

#### Security

- [ ] Security architect has reviewed the changes in this PR,
- [ ] These changes are part of a larger initiative with a separate security review, or
- [ ] There are no security aspects to these changes

### CODE REVIEW

File: conjur/logic/login_logic.py [hunk 1]
```diff
@@ -12,25 +12,28 @@
...
     def __init__(self, credentials_provider: CredentialsStoreInterface):
         self.credentials_provider = credentials_provider
 
     @classmethod
-    # pylint: disable=line-too-long,logging-fstring-interpolation
-    def get_api_key(cls, ssl_verify:bool, credential_data:CredentialsData,
-                    password:str, conjurrc:ConjurrcData) -> str :
+    # pylint: disable=logging-fstring-interpolation,try-except-raise,raise-missing-from
+    def get_api_key(cls, ssl_verification_metadata: SslVerificationMetadata, credential_data: CredentialsData,
+                    password: str, conjurrc: ConjurrcData) -> str:
```
Comment (abrahamko) [inline conjur/logic/login_logic.py:36]: If line separation is needed between params, do it between each param:

```suggestion
    def get_api_key(cls,
                    ssl_verification_metadata: SslVerificationMetadata,
                    credential_data: CredentialsData,
                    password: str,
                    conjurrc: ConjurrcData) -> str:
```

File: conjur/logic/login_logic.py [hunk 2]
```diff
@@ -39,24 +42,19 @@ def get_api_key(cls, ssl_verify:bool, credential_data:CredentialsData,
...
+            api_key = invoke_endpoint(HttpVerb.GET,
+                                      ConjurEndpoint.LOGIN,
+                                      params,
+                                      auth=(credential_data.login, password),
+                                      ssl_verification_metadata=ssl_verification_metadata).text
+        except HttpSslError:
+            if not conjurrc.cert_file and not ssl_verification_metadata.is_insecure_mode:
+                raise CertificateVerificationException
+        except Exception:
+            raise
```
Comment (abrahamko) [inline conjur/logic/login_logic.py:57]: seems redundant. no need to catch exception just to raise it, unless it is specific, but it's not the case here.

File: conjur/util/init_utils.py [hunk 3]
```diff
@@ -21,9 +24,9 @@ def validate_init_action_ssl_verification_input(ca_path, is_self_signed, ssl_ver
...
     options = [use_ca_bundle, is_self_signed, not ssl_verify]
     if sum(options) > 1:
-        raise MissingRequiredParameterException("SSL verification method can be one of three:"
-                                                "\n1. --ca-cert with '< Full path to RootCA PEM"
-                                                " file >' (recommended)"
+        raise MissingRequiredParameterException("Can't accept more than one of the following "
+                                                "arguments:"
+                                                "\n1. --ca-cert"
                                                 "\n2. --self-signed"
                                                 "\n3. --insecure (skip certificate validation)")
```
Comment (abrahamko) [inline conjur/util/init_utils.py:31]: 2 things:
1. The exception type is not correct. there's no required parameter missing, it's the opposite. there are too many parameters.
2. The exception should not contain instructions to the user, especially not in `util` functions, and more especially not instructions for cli user.
To give instructions to the user, catch this exception type in `InitController`, and write it to stdout there.
The exception should only contain data that explains what was wrong.

File: conjur/util/init_utils.py [hunk 4]
```diff
@@ -21,9 +24,9 @@ def validate_init_action_ssl_verification_input(ca_path, is_self_signed, ssl_ver
         use_ca_bundle = True
     options = [use_ca_bundle, is_self_signed, not ssl_verify]
```
Comment (abrahamko) [inline conjur/util/init_utils.py:25]: Consider using mutually exclusive in arg parse instead of this.
https://docs.python.org/3/library/argparse.html#mutual-exclusion

File: conjur/util/init_utils.py [hunk 5]
```diff
@@ -39,3 +42,21 @@ def validate_init_action_ssl_verification_input(ca_path, is_self_signed, ssl_ver
         except Exception:
             # pylint: disable=raise-missing-from
             raise InvalidFilePermissionsException(f"No read access for: {ca_path}")
+
+
+def get_ssl_verification_meta_data_from_cli_params(ca_path, is_self_signed,
+                                                   ssl_verify) -> SslVerificationMetadata:
```
Comment (abrahamko) [inline conjur/util/init_utils.py:48]: If line separation is needed between params, do it between each param:

```suggestion
def get_ssl_verification_meta_data_from_cli_params(ca_path,
                                                  is_self_signed,
                                                   ssl_verify) -> SslVerificationMetadata:
```

File: conjur/api/models/ssl_verification_mode.py [hunk 6]
```diff
@@ -12,4 +12,4 @@ class SslVerificationMode(Enum):
     WITH_TRUST_STORE = 0
     WITH_CA_BUNDLE = 1
     SELF_SIGN = 2
-    NO_SSL = 3
+    NO_SSL = 3 # TODO rename to no_cert_validation
```
Comment (abrahamko) [inline conjur/api/models/ssl_verification_mode.py:15]: do the TODO

File: conjur/cli_actions.py [hunk 7]
```diff
@@ -35,7 +35,10 @@ def handle_init_logic(
     Method that wraps the init call logic
     Initializes the client, creating the .conjurrc file
     """
-    validate_init_action_ssl_verification_input(cert, is_self_signed, ssl_verify)
+    init_utils.validate_init_action_ssl_verification_input(cert, is_self_signed, ssl_verify)
+    ssl_verification_data = init_utils.get_ssl_verification_meta_data_from_cli_params(cert,
+                                                                                      is_self_signed
```
Comment (abrahamko) [inline conjur/cli_actions.py:40]: commas not in place

File: conjur/controller/init_controller.py [hunk 8]
```diff
@@ -37,53 +39,61 @@ class InitController:
...
+        self._get_account_info_if_not_exist()
         self.write_conjurrc()
-
         sys.stdout.write("Successfully initialized the Conjur CLI\n")
 
+    def _fetch_certificate_if_needed_and_update_conjurrc(self, formatted_conjur_url):
+        mode = self.ssl_verification_data.mode
+        if mode == SslVerificationMode.WITH_CA_BUNDLE:
+            self.conjurrc_data.cert_file = self.ssl_verification_data.ca_cert_path
+        if mode == SslVerificationMode.SELF_SIGN:
```
Comment (abrahamko) [inline conjur/controller/init_controller.py:76]:
```suggestion
        elif mode == SslVerificationMode.SELF_SIGN:
```

File: conjur/controller/init_controller.py [hunk 9]
```diff
@@ -37,53 +39,61 @@ class InitController:
...
+            fetched_certificate = self._get_server_certificate(formatted_conjur_url)
+            self._write_certificate(fetched_certificate)
+        if mode in [SslVerificationMode.NO_SSL, SslVerificationMode.WITH_TRUST_STORE]:
```
Comment (abrahamko) [inline conjur/controller/init_controller.py:79]: Is `else` enough?

File: conjur/logic/login_logic.py (hunk 1 above)
Comment (mbenita-Cyberark) [reply conjur/logic/login_logic.py:36]: Done

File: conjur/logic/login_logic.py (hunk 2 above)
Comment (mbenita-Cyberark) [reply conjur/logic/login_logic.py:57]: Done

File: conjur/controller/init_controller.py [hunk 10]
```diff
@@ -94,14 +104,15 @@ def _format_conjur_url(self) -> Tuple[str, str]:
...
         """
         Validates the specified url
 
         Raises a RuntimeError in case of an invalid url format
         """
+        allow_https_only = not self.ssl_verification_data.is_insecure_mode
         valid_scheme = conjur_url.scheme == 'https' or (
-                    not allow_https_only and conjur_url.scheme == 'http')
+                not allow_https_only and conjur_url.scheme == 'http')
         if not valid_scheme:
```
Comment (abrahamko) [inline conjur/controller/init_controller.py:116]: this is very unreadable...
Maybe do this instead?
```suggestion
        # http scheme is valid only in insecure mode
        valid_scheme = conjur_url.scheme == 'https'  or self.ssl_verification_data.is_insecure_mode
        if not valid_scheme:
```

File: conjur/controller/init_controller.py [hunk 11]
```diff
@@ -189,6 +204,15 @@ def write_conjurrc(self):
                                            True)
         sys.stdout.write(f"Configuration written to {DEFAULT_CONFIG_FILE}\n\n")
 
+    @staticmethod
+    def _prompt_warning_for_self_signed_flow():
+        user_answer = input("Using self-signed certificates is not recommended and could lead to "
+                            "exposure of sensitive data.\n Continue? yes/no (Default: no): "
+                            "").strip() or "no"
```
Comment (abrahamko) [inline conjur/controller/init_controller.py:211]:
```suggestion
                            ).strip() or "no"
```

File: conjur/util/util_functions.py [hunk 12]
```diff
@@ -13,7 +13,8 @@
 import os
 
 # Internals
-from conjur.api.models import SslVerificationMetadata, SslVerificationMode
+from conjur.api.models import SslVerificationMetadata
+from conjur.api.models import SslVerificationMode
```
Comment (abrahamko) [inline conjur/util/util_functions.py:17]: Why this change is needed?

File: conjur/util/init_utils.py (hunk 5 above)
Comment (mbenita-Cyberark) [reply conjur/util/init_utils.py:48]: Done

File: conjur/util/init_utils.py (hunk 4 above)
Comment (mbenita-Cyberark) [reply conjur/util/init_utils.py:25]: Done

File: conjur/api/models/ssl_verification_mode.py (hunk 6 above)
Comment (mbenita-Cyberark) [reply conjur/api/models/ssl_verification_mode.py:15]: Done :-)

File: conjur/controller/init_controller.py (hunk 8 above)
Comment (mbenita-Cyberark) [reply conjur/controller/init_controller.py:76]: Done

File: conjur/controller/init_controller.py (hunk 9 above)
Comment (mbenita-Cyberark) [reply conjur/controller/init_controller.py:79]: changed to elif incase the enum value is None

**Review by abrahamko — APPROVED** [review]
LGTM, great work!

**Review by abrahamko — DISMISSED** [review]
LGTM, great work!

File: conjur/controller/init_controller.py [hunk 13]
```diff
@@ -189,6 +204,14 @@ def write_conjurrc(self):
                                            True)
         sys.stdout.write(f"Configuration written to {DEFAULT_CONFIG_FILE}\n\n")
 
+    @staticmethod
+    def _prompt_warning_for_self_signed_flow():
+        user_answer = input("Using self-signed certificates is not recommended and could lead to "
+                            "exposure of sensitive data.\n Continue? yes/no (Default: no): ").strip()
+        if user_answer.lower() not in VALID_CONFIRMATIONS:
+            raise OperationNotCompletedException(
+                "User decided to not work with self signed certificate")
```
Comment (mbenita-Cyberark) [inline conjur/controller/init_controller.py:213]: @shulifink

File: conjur/controller/init_controller.py [hunk 14]
```diff
@@ -137,13 +148,17 @@ def _get_server_certificate(self, conjur_url: ParseResult) -> Optional[str]:
...
         Method to fetch the account from the user
         """
         if self.conjurrc_data.conjur_account is None:
             try:
-                self.init_logic.fetch_account_from_server(self.conjurrc_data)
+                self.init_logic.fetch_account_from_server(self.conjurrc_data, self.ssl_verification_data)
+            except HttpSslError as ssl_err:
+                raise HttpSslError("SSL Error. Make sure Conjur "
+                                   "server's root certificate is trusted in this machine. "
+                                   "If this problem continues, visit docs fo more help") from ssl_err
```
Comment (mbenita-Cyberark) [inline conjur/controller/init_controller.py:161]: @shulifink

**Review by mbenita-Cyberark — COMMENTED** [review]
taging @shulifink  for comment

File: conjur/controller/init_controller.py (hunk 13 above)
Comment (shulifink) [inline conjur/controller/init_controller.py:213]:
```suggestion
                "User chose not to use self-signed certificate")
```

File: conjur/controller/init_controller.py (hunk 14 above)
Comment (shulifink) [reply conjur/controller/init_controller.py:161]: "Make sure the Conjur certificate is trusted on this machine. If this issue persists, see <where in docs??>"

Where in the docs should we refer them to? Even if it's not an exact place, we should state where about in the docs? Alternatiavley, we  could we refer them to customer support...

**Review by shulifink — CHANGES_REQUESTED** [review]

**Review by abrahamko — APPROVED** [review]
LGTM
